src/app/DeviceProxy.h

```diff
@@ -0,0 +1,132 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *  @file
+ *    This file contains definitions for DeviceProxy base class. The objects of this
+ *    class will be used by applications to interact with peer CHIP devices.
+ *    The class provides mechanism to construct, send and receive messages to and
+ *    from the corresponding CHIP devices.
+ */
+
+#pragma once
+
+#include <app/CommandSender.h>
+#include <app/DeviceControllerInteractionModelDelegate.h>
+#include <app/InteractionModelEngine.h>
+#include <app/util/CHIPDeviceCallbacksMgr.h>
+#include <lib/core/CHIPCallback.h>
+#include <lib/core/CHIPCore.h>
+#include <lib/support/DLLUtil.h>
+
+namespace chip {
+
+class DLL_EXPORT DeviceProxy
+{
+public:
+    virtual ~DeviceProxy() {}
+    DeviceProxy() {}
+
+    /**
+     *   Called when a connection is closing.
+     *   The object releases all resources associated with the connection.
+     */
+    virtual void OnConnectionExpired(SessionHandle session) = 0;
+
+    /**
+     *  Mark any open session with the device as expired.
+     */
+    virtual CHIP_ERROR Disconnect() = 0;
+
+    virtual NodeId GetDeviceId() const = 0;
+
+    virtual bool GetAddress(Inet::IPAddress & addr, uint16_t & port) const { return false; }
```

Comment:
I think it might be better to have chip::Transport::PeerAddress as output parameter so we also capture kUdp vs. kTcp and interface ID (when present), which is necessary for link-local addresses.